integration/querier_remote_read_test.go

```diff
@@ -162,118 +165,230 @@ func isEqualSampleAndHistogram(t *testing.T, expectedVector *model.Sample, histo
+				Histograms: histograms,
+			})
+
+			res, err := c.Push(series)
+			require.NoError(t, err)
+			require.Equal(t, 200, res.StatusCode)
+
+			matcher, err := labels.NewMatcher(labels.MatchEqual, "__name__", "series_1")
+			require.NoError(t, err)
+
+			q, err := remote.ToQuery(startMs, endMs, []*labels.Matcher{matcher}, &storage.SelectHints{
+				Step:  1,
+				Start: startMs,
+				End:   endMs,
+			})
+			require.NoError(t, err)
+
+			req := &prompb.ReadRequest{
+				Queries:               []*prompb.Query{q},
+				AcceptedResponseTypes: []prompb.ReadRequest_ResponseType{prompb.ReadRequest_STREAMED_XOR_CHUNKS},
+			}
+
+			data, err := proto.Marshal(req)
+			require.NoError(t, err)
+			compressed := snappy.Encode(nil, data)
+
+			// Call the remote read API endpoint with a timeout.
+			httpReqCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			httpReq, err := http.NewRequestWithContext(httpReqCtx, "POST", "http://"+querier.HTTPEndpoint()+"/prometheus/api/v1/read", bytes.NewReader(compressed))
+			require.NoError(t, err)
+			httpReq.Header.Add("Accept-Encoding", "snappy")
+			httpReq.Header.Set("X-Scope-OrgID", "user-1")
+			httpReq.Header.Set("User-Agent", "Prometheus/1.8.2")
+			httpReq.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
+
+			httpResp, err := http.DefaultClient.Do(httpReq)
+			require.NoError(t, err)
+			require.Equal(t, http.StatusOK, httpResp.StatusCode)
+
+			// Fetch streaming response
+			stream := remote.NewChunkedReader(httpResp.Body, remote.DefaultChunkedReadLimit, nil)
+
+			var results []prompb.ChunkedReadResponse
+			for {
+				var res prompb.ChunkedReadResponse
+				err := stream.NextProto(&res)
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				require.NoError(t, err)
+				results = append(results, res)
+			}
+
+			// Validate the returned remote read data
+			i := 0
+			for _, result := range results {
+				// We're only expected a single series `series_1`.
+				require.Len(t, result.ChunkedSeries, 1)
+				require.Equal(t, "series_1", result.ChunkedSeries[0].Labels[0].GetValue())
+
+				for _, rawChk := range result.ChunkedSeries[0].Chunks {
+					var enc chunkenc.Encoding
+					switch rawChk.Type {
+					case prompb.Chunk_XOR:
+						enc = chunkenc.EncXOR
+					case prompb.Chunk_HISTOGRAM:
+						enc = chunkenc.EncHistogram
+					// TODO: Once prompb.Chunk_FLOAT_HISTOGRAM gets added in https://github.com/prometheus/prometheus/pull/12085 we can switch to using that
+					case prompb.Chunk_Encoding(3):
+						enc = chunkenc.EncFloatHistogram
+					default:
+						require.Fail(t, "unrecognized chunk type")
+					}
+
+					chk, err := chunkenc.FromData(enc, rawChk.Data)
+					require.NoError(t, err)
+
+					chkItr := chk.Iterator(nil)
+					j := 0
+					for valType := chkItr.Next(); valType != chunkenc.ValNone; valType = chkItr.Next() {
+						switch valType {
+						case chunkenc.ValFloat:
+							ts, val := chkItr.At()
+							if samples != nil {
```

Comment:
Should we fail the test if we don't expect to receive floats, but receive them? (and similarly for histograms below)

Reply:
Yea, I think so--done!

Reply:
since we're testing each type separately , I think it would be simpler to set in the `testCases` structure the expected chunk type and value type, then it's a simple `require.Equal` before the switch to check that the correct type is returned. No need for if-else

Reply:
Good idea - will do!